feat(jobs): Ensure backwards compatibility

despadam opened the pull request:
<!--
Follow semantic-release guidelines for the PR title, which is used in the changelog.

Title should follow the format `<type>(<scope>): <subject>`, where
- Type is one of: build|chore|ci|docs|feat|fix|perf|refactor|revert|style|test|BREAKING CHANGE
- Scope (optional) describes the place of the change (eg a particular milestone) and is usually omitted
- subject should be a non-capitalized one-line description in present imperative tense and not ending with a period

See https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#-git-commit-guidelines for more details.
-->

## Description
<!-- Short description of the pull request -->
With this PR we make sure that `api/v4/Jobs/` is backwards compatible with `v3`.

## Motivation
<!-- Background on use case, changes needed -->
These changes allow us to deploy the new Jobs subsystem without major changes in the FE, while also accommodating the facilities that run `v3`.

## Fixes
<!-- Please provide a list of the issues fixed by this PR -->
Alongside the following changes, all Jobs tests have been updated to use `api/v4/Jobs/`.

## Changes:
<!-- Please provide a list of the changes implemented by this PR -->
NestJS annotations allow us to indicate which methods are compatible with which version(s).
- `GET/fullfacet` and `DELETE` are now compatible with both versions - no further changes were required.
- For `POST`, `PATCH`, `GET`, `GET/:id` and `GET/fullquery` we need two methods for each, one for `v3` and one for `v4`. For that, the following changes were introduced:
  - Brought back the `JobClassV3` schema
  - In the controller, added the method `mapJobClassV4toV3()` that transforms a `v4` Job to a `JobClassV3` instance
  - `GET`, `GET/:id` and `GET/fullquery` for `v3` are now using the `mapJobClassV4toV3()` method so that `JobClassV3` instances are returned and FE compatibility is ensured
  - Added `CreateJobV3MappingInterceptor` which is used before a Job creation with `v3`, to transform the dto to `v4`
  - Added `UpdateJobV3MappingInterceptor` which is used before a Job update with `v3`, to transform the dto to `v4`

## Tests included

- [x] Included for each change/fix?
- [x] Passing? <!-- Merge will not be approved unless tests pass -->

## Documentation
- [x] swagger documentation updated (required for API changes)
- [x] official documentation updated

### official documentation info
<!-- If you have updated the official documentation, please provide PR # and URL of the updated pages -->
https://github.com/sbliven/scicat-documentation/pull/3

despadam committed:
make jobs backwards compatible

sbliven commented on src/jobs/jobs.controller.ts:
```diff
@@ -766,6 +935,7 @@ export class JobsController {
     ability.can(Action.JobRead, JobClass),
   )
   @Get("/fullfacet")
+  @Version(["3", "4"])
```
Should we use `VERSION_NEUTRAL` instead of listing the versions explicitly?

despadam replied:
Let me test that just to be sure

sbliven replied:
If it works we should make it the default for all controllers IMO

nitrosx replied:
I agree

despadam replied:
I believe that the way we handle versioning falls into a special category seen in the screenshot below from https://docs.nestjs.com/techniques/versioning#multiple-versions

This makes the endpoints with unspecified or neutral version fail, so for now I will keep using `@Version(["3", "4"])` for `/fullfacet` and `/delete` as I am unable to find a workaround 

![image](https://github.com/user-attachments/assets/0e899931-648b-4d6e-ae7e-9ad2065757c0)

sbliven replied:
Ah, that's definitely not what we wanted. Maybe instead
```
app.enableVersioning({
  defaultVersion: ['3', '4']
});
```

The downside is that the swagger explorer will list all endpoints twice.

despadam replied:
I also came across this but was thinking that perhaps it should not be included in this PR as it will affect all endpoints.

sbliven replied:
A separate PR is fine. I just thought about here since you wouldn't need this line if it were already a default.

sbliven commented on src/jobs/jobs.controller.ts:
```diff
@@ -567,13 +570,64 @@ export class JobsController {
+    jobV3.emailJobInitiator = job.createdBy;
+    jobV3.type = job.type;
+    jobV3.creationTime = job.createdAt;
+    jobV3.jobStatusMessage = job.statusMessage;
```
I wonder if `jobV3.jobStatusMessage = job.statusCode` would better reflect how they were actually used. I think v3 statusMessage fields were single words (eg `jobStarted`).

despadam replied:
Sure, I will use `statusCode` instead

sbliven reviewed (approved):
This looks great. Using interceptors for the conversion is a nice design.

despadam commented:
Thanks for the review @sbliven - I will also update the documentation accordingly

sbliven commented on src/jobs/jobs.controller.ts:
```diff
@@ -588,53 +642,50 @@ export class JobsController {
-  async create(
+  async createV3(
     @Req() request: Request,
     @Body() createJobDto: CreateJobDto,
```
Testing this more I noticed a bug. The v3 endpoint is using the v4 DTO (datasets in the jobParams).

sbliven commented on src/jobs/jobs.controller.ts:
```diff
@@ -567,13 +570,64 @@ export class JobsController {
+    // Map fields from v4 to v3
+    jobV3._id = job._id;
+    jobV3.id = job.id;
+    jobV3.emailJobInitiator = job.createdBy;
```
Should be contactEmail

sbliven commented on src/jobs/jobs.controller.ts:
```diff
@@ -567,13 +570,64 @@ export class JobsController {
+    const { datasetList, ...jobParams } = job.jobParams;
+    jobV3.datasetList = datasetList as DatasetListDto[];
+    // Store v4 configVersion in v3 jobParams
+    const newJobParams = { ...jobParams, configVersion: job.configVersion };
```
v3 didn't have a config version. I suggest we omit it rather than squashing it into jobParams

despadam committed:
- re-introduce v3 DTOs
- update create/update interceptors and enforce types
- ensure contactEmail does not get overwritten
- add tests for api/v3/jobs

despadam committed:
fix linting

despadam committed:
fix linting

despadam committed:
fix failing tests

sbliven commented:
POST /api/v3/jobs returns a v4 DTO

despadam committed:
fix v3 return types

despadam commented:
> POST /api/v3/jobs returns a v4 DTO

Good catch @sbliven , I updated the return types for v3

sbliven commented:
Sorry to keep dribbling in change requests, but I was looking over the [real v3 schema](https://dacat.psi.ch/explorer/#!/Job/Job_findById) and there are still some differences.

| JobClass (v4)                          | dacat (v3)                              | JobClassV3                         | CreateJobDtoV3                      | UpdateJobDtoV3                      |
|------------------------------------|-------------------------------------|------------------------------------|-------------------------------------|-------------------------------------|
|                                    |                                     | _id (string)                       |                                     |                                     |
| id (string)                        | id (string)                         | id (string)                        |                                     |                                     |
| contactEmail (string, optional)    | emailJobInitiator (string)          | emailJobInitiator (string)         | emailJobInitiator (string)          |                                     |
| type (string)                      | type (string)                       | type (string)                      | type (string)                       |                                     |
|                                    | creationTime (string, optional)     | creationTime (string($date-time))  |                                     |                                     |
|                                    | executionTime (string, optional)    | executionTime (string($date-time)) | executionTime (string($date-time))  | executionTime (string($date-time))  |
| jobParams (object, optional)       | jobParams (object, optional)        | jobParams (object)                 | jobParams (object, optional)        |                                     |
| statusCode (string, optional)      | jobStatusMessage (string, optional) | jobStatusMessage (string)          | jobStatusMessage (string, optional) | jobStatusMessage (string, optional) |
|                                    | datasetList (object, optional)      | datasetList (array, optional)      | datasetList (array)                 |                                     |
| jobResultObject (object, optional) | jobResultObject (object, optional)  | jobResultObject (object)           | jobResultObject (object, optional)  | jobResultObject (object, optional)  |
| createdBy (string)                 | createdBy (string, optional)        |                                    |                                     |                                     |
| updatedBy (string)                 | updatedBy (string, optional)        |                                    |                                     |                                     |
| createdAt (string($date-time))     | createdAt (string, optional)        |                                    |                                     |                                     |
| updatedAt (string($date-time))     | updatedAt (string, optional)        |                                    |                                     |                                     |
| ownerGroup (string)                |                                     | ownerGroup (string)                |                                     |                                     |
| accessGroups (array)               |                                     |                                    |                                     |                                     |
| ownerUser (string)                 |                                     |                                    |                                     |                                     |
| instrumentGroup (string, optional) |                                     |                                    |                                     |                                     |
| isPublished (boolean)              |                                     |                                    |                                     |                                     |
| statusMessage (string, optional)   |                                     |                                    |                                     |                                     |
| configVersion (object)             |                                     |                                    |                                     |                                     |



- The optional->Required changes in the schema are fine. 
- The `ownerGroup` addition to JobClassV3 is a significant change. Is this needed for authorization?
- createdAt seem redundant to creationTime, but they do differ slightly in dacat jobs
- executionTime doesn't seem to be stored or set.

despadam committed:
refactor interceptors, fix job update, map executionTime correctly

despadam committed:
fix linting

sbliven commented on src/jobs/interceptors/update-job-v3-mapping.interceptor.ts:
```diff
@@ -13,14 +15,33 @@ export class UpdateJobV3MappingInterceptor implements NestInterceptor {
     const dtoV3 = request.body as UpdateJobDtoV3;
+
+    if (!dtoV3.jobStatusMessage) {
+      throw new HttpException(
```
I'm concerned about throwing an exception in the interceptor. I think we should change the DtoV3 properties to be `required: false` (even if this isn't strictly backwards compatible) rather than saying that some jobs can only be fetched over the v4 api.

sbliven replied:
(also `emailJobInitiator` above)

nitrosx replied:
I am also concern about testing for a specific version

despadam replied:
- The concern makes sense regarding `create-job-v3-mapping.interceptor.ts`, I can remove the check for `emailJobInitiator` and keep the implementation as close as possible to `contactEmail`'s in v4.
- Regarding `update-job-v3-mapping.interceptor.ts` the `jobStatusMessage` imitates v4, because there cannot be an update if `statusCode` is not provived.
- I don't see how there are now jobs that cannot be fetched over both APIs

despadam replied:
I have removed the exceptions from the interceptors, updated the v3 dtos regarding optional/required fields and I am making the tests more detailed.

despadam committed:
optional emailJobInitiator, fix jobResultObject

despadam committed:
update dtos and tests

despadam committed:
fix tests

despadam committed:
strengthen datasetList validation

despadam committed:
update JobClassV3

despadam committed:
lint

despadam committed:
use enum in controller

despadam committed:
finish tests

despadam committed:
improve test descriptions

despadam committed:
merge with release-jobs

despadam committed:
fix anonymous tests

despadam committed:
Merge branch 'release-jobs' into jobs-backwards-compatible

despadam committed:
fix testing

nitrosx commented on src/jobs/schemas/job.v3.schema.ts:
Why do we need schema v3?
Shouldn't we save every job with schema v4?

despadam replied:
All jobs are saved in the db with v4 schema. We brought back the v3 schema and dtos to enforce typing and ensure the mapping between the different versions is correct

nitrosx replied:
Apologies for being a pain.... 
Why do you need the schema v3 to validate? Wouldn't you use the dto's?

Having schema v3 in code will be confusing!!!! More so, if it is not used!!!

despadam replied:
No worries, I will see if I can make it work just by using the dtos

sbliven replied:
This is a good point @nitrosx. The `JobClassV3` is needed as a return type for the `createV3` and `updateV3` endpoints. However, I think you're right that it shouldn't have the mongoose `@Prop` annotations. It's never stored in the database, so calling it a DTO instead of a schema makes more sense.

despadam replied:
The latest commit (https://github.com/SciCatProject/scicat-backend-next/pull/1742/commits/c2b375bc57847b95e1b3243745a9c16e8e6ac24a) fixes `JobClassV3`: removed `@Schema` and `@Prop` annotations for mongoose and we are now only using `@ApiProperty` for swagger.
- Tried to use just `CreateJobDtoV3` extended with `id`, but was not possible because `CreateJobDtoV3` has readonly values but we need to be able to update the fields of the instance.
- Tried using the `openapi cli-plugin` instead of `ApiProperty` but was not able to make it work in this case. Perhaps because this class does not correspond to a db schema? Not sure, so comments and suggestions are welcome

nitrosx commented on src/jobs/jobs.service.ts:
Why do we need updated V3?

despadam replied:
To preserve the `executionTime` field from v3, we store it in `jobParams` in v4. This function makes sure that this field will be updated correctly by using `addUpdatedJobParamsField()` that was introduced in `utils.ts`. I could split this logic differently so that most of this happens in the controller and `jobs.service` and `utils` have as less changes possible. Thanks for the comments

nitrosx replied:
I would leave this function in `utils` if it is or could be used somewhere else.
If the function is specific to jobs, I would move the it within the job subsystem.
I trust your judgement

despadam replied:
It is specific to jobs, so I will move it :)

nitrosx commented on src/common/utils.ts:
```diff
@@ -837,6 +837,18 @@ export const addStatusFields = <T>(
+    jobParams: jobParams,
+  };
+};
+
```
I wonder if this function should be in the job controller or service as it is specific for the jobs?

nitrosx replied:
I wonder if this function should be in the job controller or service as it is specific for the jobs?

despadam committed:
finish tests

despadam committed:
fix utils and service for job update

despadam committed:
add tests for anonymous jobs from normal user

despadam commented:
To map the v3/v4 create/update DTOs `CreateJobV3MappingInterceptor` and `UpdateJobV3MappingInterceptor` have been introduced.

### `CreateJobV3MappingInterceptor`:

- Move `datasetList` to `jobParams`
- Move `executionTime` to `jobParams`
- If no `emailJobInitiator` is provided then assign `request.user.email` to `contactEmail` to accommodate for **anonymous** jobs
- Since v4 is using `ownerUser` and `ownerGroup` for authorization, utilize `username` that is included in the `jobParams` coming from the FE: extract `ownerUser` and `ownerGroup` of user with `findByUsername2JWTUser()`

#### Issues depending on case - for logged in users:

1. Anonymous jobs can be created only by admins:

    - Can be created if no `emailJobInitiator` is provided. Then admin will automatically be assigned to `contactEmail`. `ownerUser` and `ownerGroup` will not exist.
    - Can be created if `emailJobInitiator` is provided but `username` was not provided in `jobParams`. Then `emailJobInitiator` will automatically be assigned to `contactEmail`. `ownerUser` and `ownerGroup` will not exist. The user whose email was provided in `emailJobInitiator` will **not** have access to the job.

3. To differentiate anonymous from user-owned jobs, we take advantage of the `username` that is included in `jobParams` from the FE. Then normal users can only create jobs for themselves:

    - If no `username` is provided in `jobParams` then normal users cannnot create jobs for themselves: `ownerGroup` is required and the controller checks that the user belongs in the specified `ownerGroup`
    - Users can belong in multiple groups but the v4 dto only asks for one group: it will always be impossible to find the correct group required for authorization when using `findByUsername2JWTUser()`
    - `JobClass` in v4 inherits from `Ownable` which has `ownerGroup` as a mandatory field: makes sense for dataset authorization but why is this used for jobs?

### `UpdateJobV3MappingInterceptor`:

- Both `statusCode` and `statusMessage` get the value of `jobStatusMessage`
- To be able to match the v4 dto, if `executionTime` is provided then store it temporarily in `jobResultObject`. When performing the update, remove it from there and store it in `jobParams`.
- Removed `addStatusFields()` from `utils.ts` (it was only used by jobs) and replaced it with helper functions `addStatusFields()` and `updateJobFields()` inside `jobs.service`.  `updateJobFields()` makes sure to also update `jobResultObject` and `jobParams` accordingly in the case that `executionTime` was provided for v3

sbliven commented:
Thanks for the summary, @despadam.
1. I'm worried that some facilities might want anonymous users to submit 'public' jobs for moving data. In v3 this was possible by adding `emailJobInitiator` without logging in. In v4 both  `ownerUser` and `ownerGroup` are required in the schema; do you set these to the empty string for anonymous cases?
2. Does it make sense for `JobClass` to extend `OwnableClass`? I'm tempted to suggest extending `QueryableClass` instead, which would eliminate the `ownerGroup`, `accessGroups`, `instrumentGroup`, and `isPublished` fields. This could be a fairly big change request (eg in casl-ability). However, I'm not sure that the 'ownable' model makes sense for jobs, which are more user-focused. Plus, removing those 4 properties would make the schemas much more comparable:

| JobClass (v4)                      | dacat (v3)                          |
| ---------------------------------- | ----------------------------------- |
| id (string)                        | **id (string)**                     |
| contactEmail (string, optional)    | **emailJobInitiator (string)**      |
| type (string)                      | **type (string)**                   |
| *(createdAt)*                      | creationTime (string, optional)     |
| jobParams.executionTime (optional)           | executionTime (string, optional)    |
| jobParams (object, optional)       | jobParams (object, optional)        |
| statusCode (string, optional)      | jobStatusMessage (string, optional) |
| statusMessage (string, optional)   |                                     |
| jobParams.datasetList (optional)      | datasetList (object, optional)      |
| jobResultObject (object, optional) | jobResultObject (object, optional)  |
| createdBy (string)                 | createdBy (string, optional)        |
| updatedBy (string)                 | updatedBy (string, optional)        |
| createdAt (string($date-time))     | createdAt (string, optional)        |
| updatedAt (string($date-time))     | updatedAt (string, optional)        |
| configVersion (string, filled by scicat)   |                                     |
| ownerUser (string)                 |                                     |
| ownerGroup (string)                |                                     |
| accessGroups (array)               |                                     |
| instrumentGroup (string, optional) |                                     |
| isPublished (boolean)              |                                     |

despadam committed:
fix job.v3 schema

despadam committed:
Merge branch 'release-jobs' into jobs-backwards-compatible

despadam committed:
job by anonymous user

despadam commented:
@sbliven 
1. is now covered: Anonymous users can create anonymous jobs with the possibility to add an `emailJobInitiator`. Then these jobs can only be accessed by admins. The user whose email was provided as `emailJobInitiator` cannot access the job. In general `ownerUser` and `ownerGroup` will not exist at all for anonymous cases, something that actually puzzles me since they are required in the schema.

despadam committed:
Merge branch 'release-jobs' into jobs-backwards-compatible

despadam committed:
rename JobClassV3 to OutputJobV3Dto and fix ownerGroup compatibility logic

despadam commented:
@nitrosx @sbliven @sofyalaski 
Based on our discussion yesterday:
- renamed `JobClassV3` to `OutputJobV3Dto`
- added an empty string as `ownerGroup`, unless `username (=ownerUser)` is provided, in which case there is a check in v4 that the user should be part of the provided `ownerGroup`:
  - to make this work the solution is to keep using `findByUsername2JWTUser` and pick the last of the groups on the list. This means that when a user wants to create a job for himself in v3, the job will be automatically shared with his group (or one of them in case he belongs in multiple groups)
- Created issue #1775 to de discussed later

nitrosx reviewed (approved):
Awesome work

sofyalaski reviewed (approved):
looks good, thank you!

sbliven commented:
> to make this work the solution is to keep using findByUsername2JWTUser and pick the last of the groups on the list. This means that when a user wants to create a job for himself in v3, the job will be automatically shared with his group (or one of them in case he belongs in multiple groups)

Can you explain why it doesn't work to leave `userGroup=""` in this case? I find it unexpected that jobs get automatically shared with an arbitrary group.

sbliven commented on src/jobs/jobs.controller.ts:
```diff
@@ -362,20 +362,45 @@ export class JobsController {
+            );
+            if (jobUser === null) {
+              Logger.log(
+                "Owner user was not found, using current user instead.",
```
Should we throw an httpexception instead of just logging it?

despadam replied:
Let's address this one later. See relevant issue https://github.com/SciCatProject/scicat-backend-next/issues/1776

sofyalaski commented:
@sbliven it's because in the current controller, it is expected that the ownerGroup field includes a group and a user creating the request is part of this group for non-admin users: 
https://github.com/SciCatProject/scicat-backend-next/blob/f4a41de34480b50463460d40b5b5188aea20b43b/src/jobs/jobs.controller.ts#L399-L421

sbliven commented on src/jobs/dto/create-job.v3.dto.ts:
```diff
@@ -12,7 +12,8 @@ import { UpdateJobDtoV3 } from "./update-job.v3.dto";
 
 export class CreateJobDtoV3 extends UpdateJobDtoV3 {
```
Fields lack descriptions

sbliven commented on src/jobs/dto/output-job-v3.dto.ts:
```diff
@@ -0,0 +1,76 @@
+    required: false,
+    description: "The email of the person initiating the job request.",
+  })
+  emailJobInitiator?: string;
```
It looks like `@Optional()` is needed in order for 'optional' to show up in the swagger docs (although the '*' is correctly omitted based just on the typescript `?` type). I'm not sure if this matters.

despadam commented:
Thanks @sofyalaski,
Yes exactly, cannot have an ownerUser without an ownerGroup, because of this check. A group is required and the user should belong to that group

despadam commented:
The other solution, instead of adding arbitrary group, is assigning an empty string also to ownerUser but then that user will not have access to the job he just created

sbliven commented on src/jobs/dto/update-job.v3.dto.ts:
```diff
@@ -7,8 +7,7 @@ export class UpdateJobDtoV3 {
   readonly executionTime?: Date;
 
   @IsString()
```
Also lack descriptions

sbliven reviewed (approved):
I am satisfied with all the DTOs now, which was the most important issue to decide here.

despadam commented:
Created issue https://github.com/SciCatProject/scicat-backend-next/issues/1776 to address the remaining topics after merging this PR, specifically how we want to handle authorization with `ownerUser`/`ownerGroup`

despadam committed:
add descriptions in dtos

despadam committed:
Merge branch 'release-jobs' into jobs-backwards-compatible

despadam committed:
fix tests after merging

despadam committed:
fix test number

sbliven commented:
Looks great. This was unexpectedly arduous. Thanks for the hard work, @despadam!
